controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -635,17 +651,27 @@ private void closeConnection(Reply reply, RawClient client) {
     }
 
     private <T extends Request & WireCommand> CompletableFuture<Reply> sendRequest(RawClient connection, long requestId, T request) {
-        return connection.sendRequest(requestId, request)
+        CompletableFuture<Reply> future = Futures.futureWithTimeout(
+                () -> connection.sendRequest(requestId, request)
+                                .exceptionally(e -> {
+                                    Throwable unwrap = Exceptions.unwrap(e);
+                                    if (unwrap instanceof ConnectionFailedException || unwrap instanceof ConnectionClosedException) {
+                                        log.warn(requestId, "Connection dropped");
+                                        throw new WireCommandFailedException(request.getType(), WireCommandFailedException.Reason.ConnectionFailed);
+                                    } else {
+                                        log.error(requestId, "Request failed", e);
+                                        throw new CompletionException(e);
+                                    }
+                                }),
+                timeout.get(), executorService);
+
+        return future
                 .exceptionally(e -> {
-                    Throwable unwrap = Exceptions.unwrap(e);
-                    if (unwrap instanceof ConnectionFailedException || unwrap instanceof ConnectionClosedException) {
-                        log.warn(requestId, "Connection dropped");
+                    if (Exceptions.unwrap(e) instanceof TimeoutException) {
+                        log.warn(requestId, "Request timedout.");
                         throw new WireCommandFailedException(request.getType(), WireCommandFailedException.Reason.ConnectionFailed);
-                    } else if (unwrap instanceof AuthenticationException) {
-                        log.warn(requestId, "Authentication Exception");
```

Comment:
we should retain the auth failure logs.

Reply:
authentication exception cannot come here ever. this is an unreachable code. 

this is because auth failure will always come in as a response from segment store. 
which is handled and translated in `handleReply` method. 

I can retain it but its a code that doesnt get covered and decreases the coverage.